File: PE-common/src/main/java/nl/adaptivity/messaging/EndpointDescriptorImpl.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.messaging;

import nl.adaptivity.util.xml.SimpleXmlDeserializable;
import nl.adaptivity.xml.*;
import nl.adaptivity.xml.schema.annotations.XmlName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.xml.bind.JAXB;
import javax.xml.namespace.QName;

import java.net.URI;


/**
 * Simple pojo implementation of {@link EndpointDescriptor} that supports
 * serialization through {@link JAXB}.
 *
 * @author Paul de Vrieze
 */
@XmlDeserializer(EndpointDescriptorImpl.Factory.class)
public class EndpointDescriptorImpl implements EndpointDescriptor, XmlSerializable, SimpleXmlDeserializable {

  public static final String MY_JBI_NS = "http://adaptivity.nl/jbi";
  public static final String ELEMENTLOCALNAME = "endpointDescriptor";
  public static final QName ELEMENTNAME = new QName(MY_JBI_NS, ELEMENTLOCALNAME, "jbi");

  public static class Factory implements XmlDeserializerFactory<EndpointDescriptorImpl> {

    @Override
    public EndpointDescriptorImpl deserialize(final XmlReader reader) throws XmlException {
      return EndpointDescriptorImpl.deserialize(reader);
    }
  }

  private String mServiceLocalName;

  private String mServiceNamespace;

  private String mEndpointName;

  private URI mEndpointLocation;

  public EndpointDescriptorImpl() {}

  public EndpointDescriptorImpl(@NotNull final QName serviceName, final String endpointName, final URI endpointLocation) {
    mServiceLocalName = serviceName.getLocalPart();
    mServiceNamespace = serviceName.getNamespaceURI();
    mEndpointName = endpointName;
    mEndpointLocation = endpointLocation;
  }


  private static EndpointDescriptorImpl deserialize(final XmlReader in) throws XmlException {
    return XmlUtil.<EndpointDescriptorImpl>deserializeHelper(new EndpointDescriptorImpl(), in);
  }

  @Override
  public boolean deserializeChild(final XmlReader in) throws XmlException {
    return false; // No children
  }

  @Override
  public boolean deserializeChildText(final CharSequence elementText) {
    return false; // No child text
  }

  @Override
  public boolean deserializeAttribute(@NotNull final CharSequence attributeNamespace, @NotNull final CharSequence attributeLocalName, @NotNull final CharSequence attributeValue) {
    switch (attributeLocalName.toString()) {
      case "endpointLocation": mEndpointLocation = URI.create(attributeValue.toString()); return true;
      case "endpointName": mEndpointName = attributeValue.toString(); return true;
      case "serviceLocalName": mServiceLocalName = attributeValue.toString(); return true;
      case "serviceNS": mServiceNamespace = attributeValue.toString(); return true;
    }
    return false;
  }

  @Override
  public void onBeforeDeserializeChildren(@NotNull final XmlReader reader) {
    // do nothing
  }

  @NotNull
  @Override
  public QName getElementName() {
    return ELEMENTNAME;
  }

  @Override
  public void serialize(@NotNull final XmlWriter out) throws XmlException {
    XmlWriterUtil.smartStartTag(out, ELEMENTNAME);
    XmlWriterUtil.writeAttribute(out, "endpointLocation", mEndpointLocation!=null ? mEndpointLocation.toString() : null );
    XmlWriterUtil.writeAttribute(out, "endpointName", mEndpointName);
    XmlWriterUtil.writeAttribute(out, "serviceLocalName", mServiceLocalName);
    XmlWriterUtil.writeAttribute(out, "serviceNS", mServiceNamespace);

    XmlWriterUtil.endTag(out, ELEMENTNAME);
  }

  @Override
  public String getEndpointName() {
    return mEndpointName;
  }

  public void setEndpointName(final String endpointName) {
    mEndpointName = endpointName;
  }

  @XmlName("endpointLocation")
  String getEndpointLocationString() {
    return mEndpointLocation.toString();
  }

  void setEndpointLocationString(@NotNull final String location) {
    mEndpointLocation = URI.create(location);
  }

  @Override
  public URI getEndpointLocation() {
    return mEndpointLocation;
  }

  public void setEndpointLocation(final URI location) {
    mEndpointLocation = location;
  }

  String getServiceLocalName() {
    return mServiceLocalName;
  }

  void setServiceLocalName(final String localName) {
    mServiceLocalName = localName;
  }

  @XmlName("serviceNS")
  String getServiceNamespace() {
    return mServiceNamespace;
  }

  void setServiceNamespace(final String serviceNamespace) {
    mServiceNamespace = serviceNamespace;
  }

  @NotNull
  @Override
  public QName getServiceName() {
    return new QName(mServiceNamespace, mServiceLocalName);
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + ((mEndpointLocation == null) ? 0 : mEndpointLocation.hashCode());
    result = prime * result + ((mEndpointName == null) ? 0 : mEndpointName.hashCode());
    result = prime * result + ((mServiceLocalName == null) ? 0 : mServiceLocalName.hashCode());
    result = prime * result + ((mServiceNamespace == null) ? 0 : mServiceNamespace.hashCode());
    return result;
  }

  @Override
  public boolean equals(@Nullable final Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    final EndpointDescriptorImpl other = (EndpointDescriptorImpl) obj;
    if (mEndpointLocation == null) {
      if (other.mEndpointLocation != null)
        return false;
    } else if (!mEndpointLocation.equals(other.mEndpointLocation))
      return false;
    if (mEndpointName == null) {
      if (other.mEndpointName != null)
        return false;
    } else if (!mEndpointName.equals(other.mEndpointName))
      return false;
    if (mServiceLocalName == null) {
      if (other.mServiceLocalName != null)
        return false;
    } else if (!mServiceLocalName.equals(other.mServiceLocalName))
      return false;
    if (mServiceNamespace == null) {
      if (other.mServiceNamespace != null)
        return false;
    } else if (!mServiceNamespace.equals(other.mServiceNamespace))
      return false;
    return true;
  }

  @Override
  public boolean isSameService(final EndpointDescriptor other) {
    return mServiceNamespace.equals(other.getServiceName().getNamespaceURI()) && mServiceLocalName.equals(other.getServiceName().getLocalPart()) && mEndpointName.equals(other.getEndpointName());
  }

}

File: PE-common/common/src/main/kotlin/nl/adaptivity/messaging/EndpointDescriptorImpl.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.messaging

import nl.adaptivity.util.multiplatform.URI
import nl.adaptivity.util.multiplatform.createUri
import nl.adaptivity.util.xml.SimpleXmlDeserializable
import nl.adaptivity.xml.*


/**
 * Simple pojo implementation of [EndpointDescriptor] that supports
 * serialization through [JAXB].
 *
 * @author Paul de Vrieze
 */
@XmlDeserializer(EndpointDescriptorImpl.Factory::class)
class EndpointDescriptorImpl(serviceName: QName?,
                             override var endpointName: String?,
                             override var endpointLocation: URI?) : EndpointDescriptor, XmlSerializable, SimpleXmlDeserializable {

    internal var serviceLocalName: String? = null

    internal var serviceNamespace: String? = null

    override val elementName: QName
        get() = ELEMENTNAME

    internal var endpointLocationString: String
        get() = endpointLocation!!.toString()
        set(location) {
            endpointLocation = createUri(location)
        }

    override val serviceName: QName?
        get() = serviceLocalName?.let { QName(serviceNamespace ?: XMLConstants.NULL_NS_URI, it) }

    class Factory : XmlDeserializerFactory<EndpointDescriptorImpl> {

        override fun deserialize(reader: XmlReader): EndpointDescriptorImpl {
            return Companion.deserialize(reader)
        }
    }

    constructor():this(null, null, null)

    override fun deserializeChild(reader: XmlReader): Boolean {
        return false // No children
    }

    override fun deserializeChildText(elementText: CharSequence): Boolean {
        return false // No child text
    }

    override fun deserializeAttribute(attributeNamespace: String?,
                                      attributeLocalName: String,
                                      attributeValue: String): Boolean {
        when (attributeLocalName) {
            "endpointLocation" -> {
                endpointLocation = createUri(attributeValue)
                return true
            }
            "endpointName"     -> {
                endpointName = attributeValue
                return true
            }
            "serviceLocalName" -> {
                serviceLocalName = attributeValue
                return true
            }
            "serviceNS"        -> {
                serviceNamespace = attributeValue
                return true
            }
        }
        return false
    }

    override fun onBeforeDeserializeChildren(reader: XmlReader) {
        // do nothing
    }

    override fun serialize(out: XmlWriter) {
        out.smartStartTag(ELEMENTNAME) {
            writeAttribute("endpointLocation", endpointLocation?.toString())
            writeAttribute("endpointName", endpointName)
            writeAttribute("serviceLocalName", serviceLocalName)
            writeAttribute("serviceNS", serviceNamespace)
        }
    }

    override fun isSameService(other: EndpointDescriptor?): Boolean {
        return other!=null && serviceNamespace == other.serviceName?.getNamespaceURI() && serviceLocalName == other.serviceName?.getLocalPart() && endpointName == other.endpointName
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false

        other as EndpointDescriptorImpl

        if (endpointName != other.endpointName) return false
        if (endpointLocation != other.endpointLocation) return false
        if (serviceLocalName != other.serviceLocalName) return false
        if (serviceNamespace != other.serviceNamespace) return false

        return true
    }

    override fun hashCode(): Int {
        var result = endpointName?.hashCode() ?: 0
        result = 31 * result + (endpointLocation?.hashCode() ?: 0)
        result = 31 * result + (serviceLocalName?.hashCode() ?: 0)
        result = 31 * result + (serviceNamespace?.hashCode() ?: 0)
        return result
    }

    companion object {

        val MY_JBI_NS = "http://adaptivity.nl/jbi"
        val ELEMENTLOCALNAME = "endpointDescriptor"
        val ELEMENTNAME = QName(MY_JBI_NS,
                                ELEMENTLOCALNAME, "jbi")


        private fun deserialize(reader: XmlReader): EndpointDescriptorImpl {
            return EndpointDescriptorImpl().deserializeHelper(reader)
        }
    }

    init {
        serviceLocalName = serviceName?.localPart
        serviceNamespace = serviceName?.namespaceURI
    }

}
